### ISSUE TITLE
Scope close exception when zipping dequeues

### ISSUE THREAD

**Issue by mpilquist** [functional-streams-for-scala/fs2#1083]
```scala
object Foo extends App {
  Stream.eval(fs2.async.unboundedQueue[IO,Int]).flatMap { q =>
    val p = Stream(0,1,2).covary[IO].to(q.enqueue)
    val out = q.dequeue.zip(q.dequeue)
    out.concurrently(p)
  }.compile.drain.unsafeRunSync
}
```

This program throws:

```
[error] java.lang.IllegalStateException: Failed to close scope that is not active scope or ancestor
[error] 	at fs2.internal.Algebra$.$anonfun$compileFoldLoop$15(Algebra.scala:400)
[error] 	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:116)
[error] 	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:294)
[error] 	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:276)
[error] 	at cats.effect.internals.IOPlatform$.$anonfun$onceOnly$1(IOPlatform.scala:75)
[error] 	at cats.effect.internals.IOPlatform$.$anonfun$onceOnly$1$adapted(IOPlatform.scala:75)
[error] 	at fs2.async.Promise.$anonfun$getOrWait$9(Promise.scala:131)
[error] 	at fs2.async.Promise.$anonfun$getOrWait$9$adapted(Promise.scala:131)
[error] 	at fs2.async.Promise.$anonfun$complete$3(Promise.scala:96)
[error] 	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:140)
[error] 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
[error] 	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
[error] 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
[error] 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

The exception occurs upon the dequeue of the third element.

Another interesting observation: replacing `q.dequeue` with `Stream.repeatEval(q.dequeue1)` results in no error.

**Comment by mpilquist** [comment]
Minimization:
```scala
val s = Stream(0).scope
(s ++ s).zip(s).toList
```

**Comment by pchlupacek** [comment]
ok I am thinking of that I know where is problem, and indeed is in zip. The issue is when left opens scope, then it emits, but not yet closes scope (tail is closing scope) then right emits, opens scope (note it is now child of left scope) and emits, with cleanup on tail. Then left runs its cleanup effectivelly closing the scope of right, and then right runs its cleanup failing to do so.

**Comment by mpilquist** [comment]
One final further minimization:

```scala

  def go[F[_],O](l: Stream[F,O], r: Stream[F,O]): Pull[F,O,Unit] =
    l.pull.uncons.flatMap {
      case None => Pull.done
      case Some((hd1, tl1)) =>
        r.pull.uncons.flatMap {
          case None => Pull.done
          case Some((hd2, tl2)) =>
            Pull.output(hd1) >> Pull.output(hd2) >> go(tl1, tl2)
        }
    }

  val s = Stream(0).scope
  go(s ++ s, s).stream.toList
```

**Comment by pchlupacek** [comment]
This is what's going on...
```
Algebra.openScope(L1) >>
Algebra.output1(0-l) >>
Algebra.openScope(R1) >>  // R1 child of L1
Algebra.closeScope(L1) >> // L1 is closed, closing R1
Algebra.output(0-r) >> //ok, but will fail if (0-r) will regure resources from R1 
Algebra.closeScope(R1) // fails, R1, L1 closed already. 
```

**Comment by pchlupacek** [comment]
After more thinking about this, I think there is no easy way to fix this. Interleaving scopes are changing the stream structuru a lot. Its quite surprising it didn't manifests before. 

I am thinking in two possible lines now: 
- (A)  change signature of compileFoldLoop to be something like uncons : 
 ```scala
def compileFoldLoop(s: FreeC[Algerbra[F,O,?], Unit], scope: CompileFoldScope[F, O]): F[(Segment[F, O], CopileFoldScope[F, O],  FreeC[Algerbra[F,O,?], Unit]) = ???

// and then zip will be defined outside of current algebra something like 
// (subject to implementation of cleanups etc
Stream.eval(compileFoldLoop(left, scopeL)) flatMap { case (outL, scopeL, tailL) => 
Stream.eval(compileFoldLoop(right, scopeR)) flatMap { case (outR, scopeR, tailR) =>
   Stream.emits(outL zip outR)  ++  //... and recurse, resolve leftovers etc...
}
```
-- OR --

B) make zip to take an effect and assure the Right side will be consulted when Left side will have elements. This will require zip to take `Effect[F]` and I am not sure if that is problem or not. 

I do not have preference over which one is better. @SystemFw @mpilquist @AdamChlupacek  any other ideas // preferences?

**Comment by pchlupacek** [comment]
fixed by #1084
